llvm-external-projects/iree-dialects/include/iree-dialects/Transforms/ListenerGreedyPatternRewriteDriver.h

```diff
@@ -4,37 +4,79 @@
 #include "mlir/IR/PatternMatch.h"
 #include "mlir/Rewrite/FrozenRewritePatternSet.h"
 #include "mlir/Transforms/GreedyPatternRewriteDriver.h"
 
 // The following are iree-dialects extensions to MLIR.
 namespace mlir {
-struct GreedyRewriteConfig;
 struct RewriteListener;
 
-/// Applies the specified patterns on `op` alone while also trying to fold it,
-/// by selecting the highest benefits patterns in a greedy manner. Returns
-/// success if no more patterns can be matched. `erased` is set to true if `op`
-/// was folded away or erased as a result of becoming dead. Note: This does not
-/// apply any patterns recursively to the regions of `op`. Accepts a listener
-/// so the caller can be notified of rewrite events.
-LogicalResult applyPatternsAndFoldGreedily(
-    Operation *op, const FrozenRewritePatternSet &patterns,
-    const GreedyRewriteConfig &config, RewriteListener *listener);
+/// Rewrite ops in the given region, which must be isolated from above, by
+/// repeatedly applying the highest benefit patterns in a greedy work-list
+/// driven manner.
+///
+/// This variant may stop after a predefined number of iterations, see the
+/// alternative below to provide a specific number of iterations before stopping
+/// in absence of convergence.
+///
+/// Return success if the iterative process converged and no more patterns can
+/// be matched in the result operation regions.
+///
+/// Note: This does not apply patterns to the top-level operation itself.
+///       These methods also perform folding and simple dead-code elimination
+///       before attempting to match any of the provided patterns.
+///
+/// You may configure several aspects of this with GreedyRewriteConfig.
+LogicalResult applyPatternsTrackAndFoldGreedily(
+    RewriteListener *listener, Region &region,
+    const FrozenRewritePatternSet &patterns,
+    GreedyRewriteConfig config = GreedyRewriteConfig());
 
-/// Apply the given list of transformations to the regions of the
-/// isolated-from-above operation `root` greedily until convergence. Update
-/// Linalg operations in values of `trackedOperations` if they are replaced by
-/// other Linalg operations during the rewriting process. Tracked operations
-/// must be replaced with Linalg operations and must not be erased in the
-/// patterns.
-static inline LogicalResult applyPatternsTrackAndFoldGreedily(
-    Operation *root, RewriteListener &listener,
+/// Rewrite ops in all regions of the given op, which must be isolated from
+/// above.
+inline LogicalResult applyPatternsTrackAndFoldGreedily(
+    RewriteListener *listener, Operation *op,
     const FrozenRewritePatternSet &patterns,
     GreedyRewriteConfig config = GreedyRewriteConfig()) {
-  return applyPatternsAndFoldGreedily(root, patterns, config, &listener);
+  bool failed = false;
+  for (Region &region : op->getRegions())
+    failed |=
+        applyPatternsTrackAndFoldGreedily(listener, region, patterns, config)
+            .failed();
+  return failure(failed);
```

Comment:
Ultra-nit: something like

```
LogicalResult result = applyPatternsTrackAndFoldGreedily(listener, region, patterns, config);
failed |= result.failed();
```

will likely flow better.

Reply:
This is copied from `GreedyPatternRewriteDriver.h`. I'm going to update that it there next time I touch the file.